jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/core/proxy/WebSocketProxy.java

```diff
@@ -0,0 +1,635 @@
+                        sendCallback = Callback.from(doubleCallback, failure);
+                        break;
+
+                    default:
+                        state = State.FAILED;
+                        break;
+                }
+            }
+
+            if (sendCallback != null)
+                client.close(CloseStatus.SHUTDOWN, failure.getMessage(), sendCallback);
+            else
+                callback.failed(failure);
+        }
+
+        @Override
+        public void onClosed(CloseStatus closeStatus, Callback callback)
+        {
+            System.err.println(toString() + " onClosed(): " + closeStatus);
+            closed.countDown();
+            callback.succeeded();
+        }
+
+        public void send(Frame frame, Callback callback)
+        {
+            System.err.println(toString() + " send(): " + frame);
+
+            Callback sendCallback = callback;
+            Throwable failure = null;
+            synchronized (lock)
+            {
+                switch (state)
+                {
+                    case OPEN:
+                        if (frame.getOpCode() == OpCode.CLOSE)
+                            state = State.OCLOSED;
+                        break;
+
+                    case ICLOSED:
+                        if (frame.getOpCode() == OpCode.CLOSE)
+                        {
+                            state = State.CLOSED;
+                            sendCallback = Callback.from(callback, closeCallback);
+                        }
+                        break;
+
+                    case FAILED:
+                        failure = error;
+                        break;
+
+                    default:
+                        failure = new IllegalStateException();
+                        break;
+                }
+            }
+
+            if (failure != null)
+                callback.failed(failure);
+            else
+                client.sendFrame(frame, sendCallback, false);
+        }
+
+        @Override
+        public String toString()
+        {
+            synchronized (lock)
+            {
+                return "[Client2Proxy," + state + "] ";
+            }
+        }
+    }
+
+    class Server2Proxy implements FrameHandler
+    {
+        private CoreSession server;
+        private State state = State.NOT_OPEN;
+
+        private Callback closeCallback;
+        private Throwable error;
+
+        public BlockingQueue<Frame> receivedFrames = new BlockingArrayQueue<>();
+        protected CountDownLatch closed = new CountDownLatch(1);
+
+        public State getState()
+        {
+            synchronized (this)
+            {
+                return state;
+            }
+        }
+
+        public void connect(Callback callback)
+        {
+            System.err.println(toString() + " connect()");
+
+            Throwable failure = null;
+            synchronized (lock)
+            {
+                switch (state)
```

Comment:
likewise, `connect` is only ever called from `client2proxy.onOpen` so why does it need to grab the lock and check/change state again?

Oh I see... there are two states......